core/src/execution/datafusion/expressions/cast.rs

```diff
@@ -222,3 +263,139 @@ impl PhysicalExpr for Cast {
+    let year = values[0].parse::<i32>().unwrap_or_default();
+    let month = values.get(1).map_or(1, |m| m.parse::<u32>().unwrap_or(1));
+    let day = values.get(2).map_or(1, |d| d.parse::<u32>().unwrap_or(1));
+    let hour = values.get(3).map_or(0, |h| h.parse::<u32>().unwrap_or(0));
+    let minute = values.get(4).map_or(0, |m| m.parse::<u32>().unwrap_or(0));
+    let second = values.get(5).map_or(0, |s| s.parse::<u32>().unwrap_or(0));
+    let millisecond = values.get(6).map_or(0, |ms| ms.parse::<u32>().unwrap_or(0));
+
+    match timestamp_type {
+        "year" => parse_ymd_timestamp(year, 1, 1),
+        "month" => parse_ymd_timestamp(year, month, 1),
+        "day" => parse_ymd_timestamp(year, month, day),
+        "hour" => parse_hms_timestamp(year, month, day, hour, 0, 0, 0),
+        "minute" => parse_hms_timestamp(year, month, day, hour, minute, 0, 0),
+        "second" => parse_hms_timestamp(year, month, day, hour, minute, second, 0),
+        "millisecond" => parse_hms_timestamp(year, month, day, hour, minute, second, millisecond),
+        _ => Err(CometError::CastInvalidValue {
+            value: value.to_string(),
+            from_type: "STRING".to_string(),
+            to_type: "TIMESTAMP".to_string(),
+        }),
+    }
+}
+
+fn parse_str_to_year_timestamp(value: &str) -> CometResult<Option<i64>> {
+    get_timestamp_values(value, "year")
+}
+
+fn parse_str_to_month_timestamp(value: &str) -> CometResult<Option<i64>> {
+    get_timestamp_values(value, "month")
+}
+
+fn parse_str_to_day_timestamp(value: &str) -> CometResult<Option<i64>> {
+    get_timestamp_values(value, "day")
+}
+
+fn parse_str_to_hour_timestamp(value: &str) -> CometResult<Option<i64>> {
+    get_timestamp_values(value, "hour")
+}
+
+fn parse_str_to_minute_timestamp(value: &str) -> CometResult<Option<i64>> {
+    get_timestamp_values(value, "minute")
+}
+
+fn parse_str_to_second_timestamp(value: &str) -> CometResult<Option<i64>> {
+    get_timestamp_values(value, "second")
+}
+
+fn parse_str_to_nanosecond_timestamp(value: &str) -> CometResult<Option<i64>> {
+    get_timestamp_values(value, "millisecond")
+}
+
+fn parse_str_to_time_only_timestamp(value: &str) -> CometResult<Option<i64>> {
+    let values: Vec<_> = value.split("T").collect();
+    let time_values: Vec<_> = values[1].split(":").collect();
+    let hour = time_values[0].parse::<u32>().unwrap();
+    let minute = time_values.get(1).map_or(0, |m| m.parse::<u32>().unwrap_or(0));
+    let second = time_values.get(2).map_or(0, |s| s.parse::<u32>().unwrap_or(0));
+    let millisecond = time_values.get(3).map_or(0, |ms| ms.parse::<u32>().unwrap_or(0));
+    let datetime = chrono::Local::now().to_utc().date_naive();
+    let timestamp = datetime.and_hms_milli_opt(hour, minute, second, millisecond);
+    Ok(Some(timestamp.unwrap().and_utc().timestamp_millis()))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn parse_timestamp_test() {
```

Comment:
It is good to have some tests here in Rust, but could you also enable the following Scala test in `CometCastSuite`:

```scala
  ignore("cast string to timestamp") {
    val values = Seq("2020-01-01T12:34:56.123456", "T2") ++ generateStrings(timestampPattern, 8)
    castTest(values.toDF("a"), DataTypes.DoubleType)
  }
```

Reply:
Also you will need to fix my copy and paste error in this test and change `DoubleType` to `TimestampType` 🤦